hi-ml-histopathology/src/histopathology/preprocessing/create_panda_tiles_dataset.py

```diff
@@ -60,109 +79,114 @@ def save_image(array_chw: np.ndarray, path: Path) -> PIL.Image:
     mask_tiles = mask_tiles[selected]
     tile_locations = tile_locations[selected]
     occupancies = occupancies[selected]
 
-    abs_tile_locations = (sample['scale'] * tile_locations + sample['location']).astype(int)
+    top, left = sample['location']
+    offset = np.array((left, top))
+
+    abs_tile_locations = (sample['scale'] * tile_locations + offset).astype(int)
+    tile_size_scaled = int(tile_size * sample['scale'])
+    tile_boxes = [Box(x, y, tile_size_scaled, tile_size_scaled) for x, y in abs_tile_locations]
 
-    return image_tiles, mask_tiles, abs_tile_locations, occupancies, n_discarded
+    return image_tiles, mask_tiles, tile_boxes, occupancies, num_discarded
 
 
 # TODO refactor this to separate metadata identification from saving. We might want the metadata
 # even if the saving fails
 def save_tile(sample: dict, image_tile: np.ndarray, mask_tile: np.ndarray,
-              tile_location: Sequence[int], output_dir: Path) -> dict:
-    slide_id = sample['image_id']
-    descriptor = get_tile_descriptor(tile_location)
-    image_tile_filename = f"train_images/{descriptor}.png"
-    mask_tile_filename = f"train_label_masks/{descriptor}_mask.png"
+              tile_box: Box, output_dir: Path) -> dict:
+    slide_id = sample[SlideKey.SLIDE_ID]
+    tile_id = get_tile_id(slide_id, tile_box)
+    image_tile_filename = f"train_images/{tile_id}.png"
+    mask_tile_filename = f"train_label_masks/{tile_id}_mask.png"
 
     save_image(image_tile, output_dir / image_tile_filename)
     save_image(mask_tile, output_dir / mask_tile_filename)
 
+    slide_metadata = sample['metadata']
     tile_metadata = {
-        'slide_id': slide_id,
-        'tile_id': get_tile_id(slide_id, tile_location),
-        'image': image_tile_filename,
-        'mask': mask_tile_filename,
-        'tile_x': tile_location[0],
-        'tile_y': tile_location[1],
-        'data_provider': sample['data_provider'],
-        'slide_isup_grade': sample['isup_grade'],
-        'slide_gleason_score': sample['gleason_score'],
+        TileKey.SLIDE_ID.value: slide_id,
+        TileKey.TILE_ID.value: tile_id,
+        TileKey.IMAGE.value: image_tile_filename,
+        TileKey.MASK.value: mask_tile_filename,
+        TileKey.TILE_LEFT.value: tile_box.x,
+        TileKey.TILE_TOP.value: tile_box.y,
+        TileKey.TILE_RIGHT.value: tile_box.x + tile_box.w,
+        TileKey.TILE_BOTTOM.value: tile_box.y + tile_box.h,
+        'data_provider': slide_metadata['data_provider'],
+        'slide_isup_grade': slide_metadata['isup_grade'],
+        'slide_gleason_score': slide_metadata['gleason_score'],
     }
 
     return tile_metadata
 
 
 def process_slide(sample: dict, level: int, margin: int, tile_size: int, occupancy_threshold: int,
-                  output_dir: Path, tile_progress: bool = False) -> None:
-    slide_id = sample['image_id']
+                  output_dir: Path, tile_progress: bool = False, filter_slide: str = '') -> None:
```

Comment:
not sure I understand what filter_slide is used for? Should this be a list?

Reply:
I used it for debugging. It's useful to generate a dataset of a single slide. I'll add a help message.